### DISCUSSION TITLE
Does `JsonSerializer.Deserialize<JsonNode>(...)` cause lazy parse?

### DISCUSSION THREAD

**Discussion by GF-Huang** [dotnet/runtime#82345] [Q&A]
The docs of `JsonSerializer` seems not detail how it works. Consider the follow:

```cs
namespace MyLibrary {
    public class Model {
        public int Number => _node[nameof(Number)].GetValue<int>();    

        private readonly JsonNode _node;

        internal Model(JsonNode node) {
            _node = node;
        }
    }

    public class Api {
        public Model GetModel(...) {
            var node = JsonSerializer.Deserialize<JsonNode>(...);
            return new Model(node);
        }
    }
}

...

namespace LibraryUser {
    public class Program {
        public static void Main() {
            var model = Api.GetModel(...);
            Console.WriteLine(model.Number); // First time access, will this cause lazy parse json here?
            ...
            Console.WriteLine(model.Number); // Second time access, will this cause parse json here again?
        }
    }
}
```

Maybe you will ask, why not deserialize into `Model` directly instead of `JsonNode`? 

Because in a real environment, a json object may have dozens or even hundreds of fields, but not all of these fields appear at the same time. To define all fields as properties (rather than `get` accessors) in the `Model`  is an unnecessary waste of memory.

**Comment by Clockwork-Muse** [comment]
> Because in a real environment, a json object may have dozens or even hundreds of fields, but not all of these fields appear at the same time.

.... you have a completely flat data model?  That seems unusual.  If you instead split the model up into multiple domain types your deserialized size could be smaller than a flatter model due to nested properties not being deserialized in some cases.

> To define all fields as properties (rather than `get` accessors) in the `Model` is an unnecessary waste of memory.

You're retaining the JSON structure.  There's a break-even point, sure, but it's probably around 25-50% of the fields being filled.  If your data is consistently mostly filled you're losing just due to object references.  If they're consistently mostly empty your data model is likely too sparse.

Note that this only applies if you're receiving lots of instances of models in a relatively short period of time, like for a web api.  If these are config values in a client application, the savings are mostly irrelevant.

> // Second time access, will this cause parse json here again?

The answer would be an internal implementation detail, and not something you should rely on.

**Reply by GF-Huang** [reply]
Not flat, but fields depends on a specific enum field value. Maybe I should write a custom converter for deserializing into different derived classes.

**Comment by Clockwork-Muse** [comment]
[Polymorphic serialization is already supported](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/polymorphism?pivots=dotnet-7-0).

**Reply by GF-Huang** [reply]
Oh, thanks point it out.

**Reply by GF-Huang** [reply]
There's a bit crappy is the type discriminator must be the first property in the json. 

![image](https://user-images.githubusercontent.com/4510984/219883809-7702261b-3c21-48da-9d05-66170d7b5b17.png)

If I move `ActivityType` to first, it works well, but usually I can't control properties order if it is a Web API.

**Reply by Clockwork-Muse** [reply]
Yes, see #72604.
